src/main/java/com/apicatalog/jsonld/flattening/NodeMapBuilder.java

```diff
@@ -123,296 +167,347 @@ public NodeMap build() throws JsonLdError {
+        // 6.12.
+        for (String property : Utils.index(elementObject.keySet(), true)) {
 
-                if (nodeMap.contains(activeGraph, id, Keywords.INDEX)) {
-                    throw new JsonLdError(JsonLdErrorCode.CONFLICTING_INDEXES);
-                }
+            final JsonValue value = elementObject.get(property);
 
-                nodeMap.set(activeGraph, id, Keywords.INDEX, elementObject.get(Keywords.INDEX));
-                elementObject.remove(Keywords.INDEX);
+            // ignore invalid expanded values - see expansion test #122
+            if (value == null || !ValueType.ARRAY.equals(value.getValueType()) && !ValueType.OBJECT.equals(value.getValueType())) {
+                continue;
             }
 
-            // 6.9.
-            if (elementObject.containsKey(Keywords.REVERSE)) {
+            // 6.12.1.
+            if (BlankNode.hasPrefix(property)) {
+                property = nodeMap.createIdentifier(property);
+            }
 
-                // 6.9.1.
-                Map<String, JsonValue> referenced = new LinkedHashMap<>();
-                referenced.put(Keywords.ID, JsonProvider.instance().createValue(id));
+            // 6.12.2.
+            if (!nodeMap.contains(activeGraph, id, property)) {
+                nodeMap.set(activeGraph, id, property, JsonValue.EMPTY_JSON_ARRAY);
+            }
 
-                // 6.9.2.
-                JsonValue reverseMap = elementObject.get(Keywords.REVERSE);
+            // 6.12.3.
+            NodeMapBuilder
+                    .with((JsonStructure) value, nodeMap)
+                    .activeGraph(activeGraph)
+                    .activeSubject(id)
+                    .activeProperty(property)
+                    .build();
+        }
+    }
 
-                // 6.9.3.
-                for (Entry<String, JsonValue> entry : reverseMap.asJsonObject().entrySet()) {
+    private static boolean noneMatch(JsonArray activePropertyValue, JsonStructure reference) {
 
-                    // 6.9.3.1.
-                    for (JsonValue value : entry.getValue().asJsonArray()) {
+        if (activePropertyValue.isEmpty()) {
+            return true;
+        }
 
-                        // 6.9.3.1.1.
-                        NodeMapBuilder
-                            .with((JsonStructure)value, nodeMap)
-                            .activeGraph(activeGraph)
-                            .referencedNode(referenced)
-                            .activeProperty(entry.getKey())
-                            .build();
-                    }
-                }
+        int referenceHashCode = reference.hashCode();
 
-                // 6.9.4.
-                elementObject.remove(Keywords.REVERSE);
+        for (int i = 0, activePropertyValueSize = activePropertyValue.size(); i < activePropertyValueSize; i++) {
+            JsonValue e = activePropertyValue.get(i);
+            if (referenceHashCode == e.hashCode()) {
```

Comment:
When checking that the `reference` object is not in the `activePropertValue` JsonArray the majority of the time was spent in the .equals method. This was taking up a significant amount of the total time. 

Since the hashCode is cached within each object, checking if the hashCode is the same dramatically reduces the time it takes to check for the existence of the `reference` object. 

I also noticed that the `reference` object is never present when converting the benchmark file. So I presume that this code is meant as some kind of edge case handling.